File: with-text-sample/pdfkit-repro/test-write-options.m
#import <Foundation/Foundation.h>
#import <PDFKit/PDFKit.h>

static void analyze(NSString *path) {
    NSData *data = [NSData dataWithContentsOfFile:path];
    NSString *text = [[NSString alloc] initWithData:data encoding:NSISOLatin1StringEncoding];
    NSUInteger s4 = 0, p1 = 0;
    NSRange r = NSMakeRange(0, text.length);
    while ((r = [text rangeOfString:@"ShadingType 4" options:0 range:r]).location != NSNotFound) { s4++; r.location++; r.length--; }
    r = NSMakeRange(0, text.length);
    while ((r = [text rangeOfString:@"PatternType 1" options:0 range:r]).location != NSNotFound) { p1++; r.location++; r.length--; }
    printf("%s  ver=%.8s  S4=%4lu  P1=%4lu  bytes=%lu\n",
           path.lastPathComponent.UTF8String, data.bytes, s4, p1, (unsigned long)data.length);
}

static BOOL writeDoc(PDFDocument *doc, NSString *path, NSDictionary *options, BOOL useData) {
    [[NSFileManager defaultManager] removeItemAtPath:path error:nil];
    if (useData) {
        NSData *data = options ? [doc dataRepresentationWithOptions:options] : [doc dataRepresentation];
        if (!data) return NO;
        return [data writeToFile:path atomically:YES];
    }
    return options ? [doc writeToURL:[NSURL fileURLWithPath:path] withOptions:options]
                   : [doc writeToURL:[NSURL fileURLWithPath:path]];
}

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc < 3) { fprintf(stderr, "usage: %s INPUT.pdf OUTDIR\n", argv[0]); return 1; }
        NSString *input = @(argv[1]);
        NSString *outDir = @(argv[2]);
        PDFDocument *doc = [[PDFDocument alloc] initWithURL:[NSURL fileURLWithPath:input]];
        if (!doc) { fprintf(stderr, "cannot open input\n"); return 1; }
        printf("input  major=%ld minor=%ld\n", (long)doc.majorVersion, (long)doc.minorVersion);
        analyze(input);

        NSArray<NSDictionary *> *cases = @[
            @{@"name": @"write-default", @"options": [NSNull null], @"data": @NO},
            @{@"name": @"dataRepresentation", @"options": [NSNull null], @"data": @YES},
            @{@"name": @"burnInAnnotations", @"options": @{PDFDocumentBurnInAnnotationsOption: @YES}, @"data": @NO},
            @{@"name": @"saveWithCorePDFLayout", @"options": @{@"SaveWithCorePDFLayout": @YES}, @"data": @NO},
            @{@"name": @"useAppendMode", @"options": @{@"UseAppendMode": @YES}, @"data": @NO},
            @{@"name": @"corepdf-plus-append", @"options": @{@"SaveWithCorePDFLayout": @YES, @"UseAppendMode": @YES}, @"data": @NO},
        ];

        for (NSDictionary *c in cases) {
            NSString *name = c[@"name"];
            NSString *out = [outDir stringByAppendingPathComponent:[name stringByAppendingString:@".pdf"]];
            NSDictionary *opts = c[@"options"] == [NSNull null] ? nil : c[@"options"];
            BOOL useData = [c[@"data"] boolValue];
            if (!writeDoc(doc, out, opts, useData)) {
                printf("%s: FAILED\n", name.UTF8String);
                continue;
            }
            analyze(out);
        }
    }
    return 0;
}

File: with-text-sample/pdfkit-repro/test-write-options.swift
import AppKit
import Foundation
import PDFKit

func analyze(_ path: String) -> String {
    let data = (try? Data(contentsOf: URL(fileURLWithPath: path))) ?? Data()
    let text = String(data: data, encoding: .isoLatin1) ?? ""
    return String(format: "%@  ver=%@  S4=%4d  P1=%4d  bytes=%d",
                  (path as NSString).lastPathComponent,
                  String(text.prefix(8)),
                  text.components(separatedBy: "ShadingType 4").count - 1,
                  text.components(separatedBy: "PatternType 1").count - 1,
                  data.count)
}

let input = CommandLine.arguments[1]
let outDir = CommandLine.arguments[2]
guard let doc = PDFDocument(url: URL(fileURLWithPath: input)) else {
    fputs("cannot open input\n", stderr)
    exit(1)
}

print("input  major=\(doc.majorVersion) minor=\(doc.minorVersion)")
print(analyze(input))

struct Case {
    let name: String
    let options: [PDFDocumentWriteOption: Any]?
    let useDataRepresentation: Bool
}

// Undocumented keys found in PDFKit.tbd (private SPI — may be ignored)
let saveWithCorePDFLayout = "SaveWithCorePDFLayout" as PDFDocumentWriteOption
let useAppendMode = "UseAppendMode" as PDFDocumentWriteOption

let cases: [Case] = [
    Case(name: "write-default", options: nil, useDataRepresentation: false),
    Case(name: "dataRepresentation", options: nil, useDataRepresentation: true),
    Case(name: "burnInAnnotations", options: [.burnInAnnotations: true], useDataRepresentation: false),
    Case(name: "saveWithCorePDFLayout", options: [saveWithCorePDFLayout: true], useDataRepresentation: false),
    Case(name: "useAppendMode", options: [useAppendMode: true], useDataRepresentation: false),
    Case(name: "corepdf-plus-append", options: [saveWithCorePDFLayout: true, useAppendMode: true], useDataRepresentation: false),
]

for testCase in cases {
    let out = (outDir as NSString).appendingPathComponent("\(testCase.name).pdf")
    try? FileManager.default.removeItem(atPath: out)

    let ok: Bool
    if testCase.useDataRepresentation {
        if let data = doc.dataRepresentation() {
            ok = (try? data.write(to: URL(fileURLWithPath: out))) != nil
        } else {
            ok = false
        }
    } else if let options = testCase.options {
        ok = doc.write(to: URL(fileURLWithPath: out), withOptions: options)
    } else {
        ok = doc.write(to: URL(fileURLWithPath: out))
    }

    print(ok ? analyze(out) : "\(testCase.name): FAILED")
}